lib/core/include/scipp/core/spatial_transforms.h

```diff
@@ -0,0 +1,194 @@
+operator*(const ScalingTransform &lhs, const TranslationTransform &rhs) {
+  return lhs.matrix() * Eigen::Translation<double, 3>(rhs.vector());
+};
+
+[[nodiscard]] inline Eigen::Affine3d operator*(const ScalingTransform &lhs,
+                                               const Eigen::Affine3d &rhs) {
+  return lhs.matrix() * rhs;
+};
+
+[[nodiscard]] inline Eigen::Vector3d operator*(const ScalingTransform &lhs,
+                                               const Eigen::Vector3d &rhs) {
+  return lhs.matrix() * rhs;
+};
+
+[[nodiscard]] inline Eigen::Affine3d operator*(const TranslationTransform &lhs,
+                                               const ScalingTransform &rhs) {
+  return Eigen::Translation<double, 3>(lhs.vector()) * rhs.matrix();
+};
+
+[[nodiscard]] inline Eigen::Affine3d operator*(const TranslationTransform &lhs,
+                                               const RotationTransform &rhs) {
+  return Eigen::Translation<double, 3>(lhs.vector()) * rhs.matrix();
+};
+
+[[nodiscard]] inline Eigen::Affine3d operator*(const TranslationTransform &lhs,
+                                               const Eigen::Matrix3d &rhs) {
+  return Eigen::Translation<double, 3>(lhs.vector()) * rhs;
+};
+
+[[nodiscard]] inline TranslationTransform
+operator*(const TranslationTransform &lhs, const TranslationTransform &rhs) {
+  return TranslationTransform(lhs.vector() + rhs.vector());
+};
+
+[[nodiscard]] inline Eigen::Affine3d operator*(const TranslationTransform &lhs,
+                                               const Eigen::Affine3d &rhs) {
+  return Eigen::Translation<double, 3>(lhs.vector()) * rhs;
+};
+
+[[nodiscard]] inline Eigen::Vector3d operator*(const TranslationTransform &lhs,
+                                               const Eigen::Vector3d &rhs) {
+  return lhs.vector() + rhs;
+};
+
+[[nodiscard]] inline Eigen::Affine3d
+operator*(const Eigen::Affine3d &lhs, const TranslationTransform &rhs) {
+  return lhs * Eigen::Translation<double, 3>(rhs.vector());
+}
+
+[[nodiscard]] inline Eigen::Affine3d operator*(const Eigen::Affine3d &lhs,
+                                               const RotationTransform &rhs) {
+  return Eigen::Affine3d(lhs * rhs.matrix());
+}
+
+[[nodiscard]] inline Eigen::Affine3d operator*(const Eigen::Affine3d &lhs,
+                                               const ScalingTransform &rhs) {
+  return Eigen::Affine3d(lhs * rhs.matrix());
+}
+
+[[nodiscard]] inline Eigen::Affine3d
+operator*(const Eigen::Matrix3d &lhs, const TranslationTransform &rhs) {
+  return lhs * Eigen::Translation<double, 3>(rhs.vector());
+}
+
+[[nodiscard]] inline Eigen::Matrix3d operator*(const Eigen::Matrix3d &lhs,
+                                               const RotationTransform &rhs) {
+  return lhs * rhs.matrix();
+}
+
+[[nodiscard]] inline Eigen::Matrix3d operator*(const Eigen::Matrix3d &lhs,
+                                               const ScalingTransform &rhs) {
+  return lhs * rhs.matrix();
+}
```

Comment:
There is a ton of duplication above. Can you avoid it by using a template (or a small number of templates)?